Add seed file support to Browsertrix backend

tw4l opened the pull request:
Fixes #2673 

Changes in this PR:

- Adds a new `file_uploads.py` module and corresponding `/files` API prefix with methods/endpoints for uploading, GETing, and deleting seed files (can be extended to other types of files moving forward)
- Seed files are supported via `CrawlConfig.config.seedFileId` on POST and PATCH endpoints. This seedFileId is replaced by a presigned url when passed to the crawler by the operator
- Seed files are read when first uploaded to calculate `firstSeed` and `seedCount` and store them in the database, and this is copied into the workflow and crawl documents when they are created.
- Logic is added to store `firstSeed` and `seedCount` for other workflows as well, and a migration added to backfill data, to maintain consistency and fix some of the pymongo aggregations that previously assumed all workflows would have at least one `Seed` object in `CrawlConfig.seeds`
- Seed file and thumbnail storage stats are added to org stats
- Seed file and thumbnail uploads first check that the org's storage quota has not been exceeded and return a 400 if so
- A cron background job (run weekly each Sunday at midnight by default, but configurable) is added to look for seed files at least x minutes old (1440 minutes, or 1 day, by default, but configurable) that are not in use in any workflows, and to delete them when they are found. The backend pods will ensure this k8s batch job exists when starting up and create it if it does not already exist. A database entry for each run of the job is created in the operator on job completion so that it'll appear in the `/jobs` API endpoints, but retrying of this type of regularly scheduled background job is not supported as we don't want to accidentally create multiple competing scheduled jobs.
- Adds a `min_seed_file_crawler_image` value to the Helm chart that is checked before creating a crawl from a workflow if set. If a workflow cannot be run, return the detail of the exception in `CrawlConfigAddedResponse.errorDetail` so that we can display the reason in the frontend

~~## Todo~~

~~- Modify test chart crawler release back to latest once crawler with seed file support is properly released~~

tw4l committed:
Add new file uplooads module with PUT and DELETE endpoints

tw4l committed:
Add temporary pylint disable comment

tw4l committed:
Add file upload tests

tw4l committed:
Add storage quota check

tw4l committed:
Fixups

tw4l committed:
Fix linting

tw4l committed:
WIP: Add seedFileId to workflow POST and PATCH

tw4l committed:
Pass seed file to crawler, add tests

tw4l committed:
Remove extra cruft

tw4l committed:
Use empty str instead of None for seed_file_url

tw4l committed:
Fix crawlconfig test expected value

tw4l committed:
Use all lowercase for upload type

tw4l committed:
Resolve internal path for presigned seed file url

tw4l committed:
Fix search values to not assume all workflows have seeds

tw4l committed:
Fix duplicate code lint warning for initing same class

tw4l committed:
Add slashes to end of page urls in test

tw4l committed:
Calculate firstSeed and seedCount for seed files on upload

tw4l committed:
Store seedCount and firstSeed in db for configs and crawls

Includes migration to calculate values for existing documents

tw4l committed:
Include seed file and thumbnail storage in org stats

Thumbnail and seed file org storage is now tracked per-org and a
migration is added to recalculate storage for existing orgs.

tw4l committed:
Modify migration to be more efficient

tw4l committed:
Add weekly cleanup cron job

When backend container starts up, ensure any regular cron cleanup
jobs exist. For now this is just seed file cleanup, but could be
easily extended.

tw4l committed:
Throw exception on job failure so job is marked as unsuccessful

tw4l commented on backend/btrixcloud/crawlconfigs.py:
```diff
@@ -456,6 +496,35 @@ async def update_crawl_config(
+            )
+
+        if update.config and update.config.seeds:
+            if update.config.seedFileId:
```
```suggestion
            if update.config.seedFileId or (
                orig_crawl_config.config.seedFileId and update.config.seedFileId is not None
            ):
```

tw4l committed:
Add additional check on crawlconfig update

tw4l commented on backend/btrixcloud/models.py:
```diff
@@ -1253,6 +1258,73 @@ def get_image_file(
+    seedCount: Optional[int] = None
+
+    async def get_file_out(self, org, storage_ops) -> UserUploadFileOut:
+        """Get ImageFileOut with new presigned url"""
```
```suggestion
        """Get UserUploadFileOut with new presigned url"""
```

tw4l committed:
Fix docstring

tw4l committed:
Enforce min_seed_file_crawler_image

Return error detail in errorDetail field of workflow added
response if the workflow can't be run so that we can display a
message in the frontend.

tw4l committed:
Add exception handler for reading cleanup cron job

tw4l committed:
Set test crawlerChannel to 1.7.0-beta.0

tw4l committed:
Fix some failing tests

tw4l committed:
Add back aggregation stage to unset config

tw4l committed:
Fix expected workflow name

tw4l committed:
Ensure seed file and thumbnail storage is added to org.bytesStored

tw4l committed:
Last test fix

tw4l committed:
Add list endpoint for files

tw4l committed:
Fix list_user_files default sortBy

tw4l commented:
Some of the all crawls tests toward the end of the test run `in test_uploads.py` are periodically returning 502, I'm not quite sure what's happening there but will keep looking into it. Otherwise this is now ready for review.

SuaYoo commented:
It doesn't appear that `path` is a real path that users can access (e.g. `path: "http://browsertrix-cloud-frontend.default/data/...`), can we update this so that users can view their uploaded file?

tw4l committed:
Add missing Union import

tw4l commented:
> It doesn't appear that `path` is a real path that users can access (e.g. `path: "http://browsertrix-cloud-frontend.default/data/...`), can we update this so that users can view their uploaded file?

This is particular to development, and the URL produced should be consistent with crawl WACZs in that same environment. In production we use proper S3-compatible services that return full presigned URLs. Locally with Minio (e.g. in development), the presigned URL that is returned is just `/data/...` so we have some logic in the backend to add the frontend origin as a prefix to this URL so that the download still works in the browser. The frontend alias defaults to `http://browsertrix-cloud-frontend.default` but can be set to a different value with `frontend_alias` in the Helm chart values file.

tw4l committed:
Check to make sure seed file not in use before deleting

Adds an additional check in case seed files are in use in multiple
workflows, e.g. in the case of workflow duplication or API users

tw4l committed:
Fix migration

tw4l committed:
Don't reuse match_query variable name

tw4l committed:
Add test for deleting in-use seed file

SuaYoo reviewed (approved):
Reviewing functionality of API changes against `frontend-upload-seed-url-list`, working well! I didn't review code or other functionality.

tw4l committed:
Make seed file cleanup job cron schedule configurable

tw4l committed:
Add nightly test for seed file cleanup jobs

tw4l committed:
TEMP FOR TESTING: Run nightly tests on this branch

tw4l committed:
Fix handling of seed cleanup jobs

tw4l committed:
Fixup

tw4l committed:
Remove ops-configs volume from background cron job yaml

tw4l committed:
Try running job in default namespace

tw4l commented:
~~Converted to draft while making sure the cron job to clean up unused seed files is working as expected. Getting close but there seems to be a permissions issue preventing the batch job from being created in the default k8s namespace:~~

```
future: <Task finished name='Task-4' coro=<BackgroundJobOps.ensure_cron_cleanup_jobs_exist() done, defined at /app/btrixcloud/background_jobs.py:478> exception=FailToCreateError([ApiException()])>
Traceback (most recent call last):
  File "/app/btrixcloud/background_jobs.py", line 480, in ensure_cron_cleanup_jobs_exist
    await self.crawl_manager.ensure_cleanup_seed_file_cron_job_exists()
  File "/app/btrixcloud/crawlmanager.py", line 257, in ensure_cleanup_seed_file_cron_job_exists
    await self.create_from_yaml(data, namespace=DEFAULT_NAMESPACE)
  File "/app/btrixcloud/k8sapi.py", line 190, in create_from_yaml
    created = await create_from_dict(
              ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/kubernetes_asyncio/utils/create_from_yaml.py", line 148, in create_from_dict
    raise FailToCreateError(api_exceptions)
kubernetes_asyncio.utils.create_from_yaml.FailToCreateError: Error from server (Forbidden): {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"cronjobs.batch is forbidden: User \"system:serviceaccount:default:default\" cannot create resource \"cronjobs\" in API group \"batch\" in the namespace \"default\"","reason":"Forbidden","details":{"group":"batch","kind":"cronjobs"},"code":403}
```

~~The job was previously in the `crawlers` namespace, but that meant it didn't have access to secrets it needed e.g. for database access.~~

~~Will continue to work on this tomorrow unless @ikreymer beats me to the punch~~

Edit: resolved

tw4l commented on .github/workflows/k3d-nightly-ci.yaml:
```diff
@@ -1,12 +1,18 @@
 on:
+  push:
+    branches:
+      - issue-2673-seed-file-backend-support
```
```suggestion
```

To remove before merging - added to test seed file cleanup jobs in PR

ikreymer replied:
Can remove this now - can still trigger nightly run on demand via GitHub

ikreymer commented on backend/btrixcloud/file_uploads.py:
```diff
@@ -0,0 +1,423 @@
+
+        print(f"{upload_type} stream upload complete", flush=True)
+
+        file_obj = file_prep.get_image_file(org.storage)
```
We should rename ImageFile / ImageFIlePreparer to something like UserFile / UserFilePreparer if its now used for more than just images, otherwise can be very confusing..

tw4l replied:
Took a stab at it in https://github.com/webrecorder/browsertrix/pull/2710/commits/ea304912b74dcc828749317bdba8e832129f2614.

This is less confusing in that way, but there are still essentially two sets of models for user-uploaded files, which may be confusing in its own right, because collection thumbnails are still stored on a different document rather than in the new `files` mongo collection.

Open to suggestions, or we can treat this as a follow-up. I left comments on the old set of models to help explain in the meantime.

ikreymer committed:
rolebinding: allow creating cronjobs in default namespace

tw4l committed:
Add back ops-configs volume

tw4l committed:
Cleanup cronjobs in default namespace on reset

tw4l committed:
Fix cleanup cron job and tests for it

Also makes delta for how old unused files need to be before
getting cleaned up configurable via helm chart so that we
can set it to 1 minute for nightly tests.

tw4l committed:
Rename ImageFile -> UserFile

This is probably a step forward but may still be confusing until
we rework collection thumbnails to be stored in the new files db
collection as well and rationalize into one set of models.

tw4l commented:
@ikreymer Seed file cleanup is now working, tested locally, passing in CI, and more configurable via helm chart values. Ready for your eyes again.

tw4l committed:
Extend seed file cleanup test to include file not cleaned up

tw4l committed:
Remove extra space

ikreymer commented on backend/btrixcloud/crawlconfigs.py:
```diff
@@ -1044,6 +1115,25 @@ async def run_now_internal(
+            seed_file_out = await self.file_ops.get_file_out(
+                crawlconfig.config.seedFileId, org
+            )
+            seed_file_url = self.storage_ops.resolve_internal_access_path(
```
Probably ok, but this might be not needed, since it seems like get_file_out() calls, which calls UserUploadFile.get_file_out() which also does resolve_internal_access_path().
I guess as an optimization, could add a UserUploadFile.get_internal_access_path() to just return the presigned URL there..

ikreymer replied:
Or, really can just remove this and use seed_file_out.path directly

tw4l replied:
Ah yeah, I think I added the `resolve_internal_access_path `UserUploadFile.get_file_out()` after adding this, but it is redundant at this point so will remove it.

tw4l committed:
Remove reundant resolve_internal_access_path call

ikreymer committed:
Update .github/workflows/k3d-nightly-ci.yaml

Co-authored-by: Tessa Walsh <tessa@bitarchivist.net>

ikreymer committed:
- add resolve_relative_access_path() which will resolve relative urls either to current origin or to internal origin, if no headers provided
- use resolve_relative_access_path() for both file uploads and collection thumbnails

ikreymer committed:
make useruploadfile derive from userfile

ikreymer committed:
fix namespace on seed cleanup cronjob check

ikreymer committed:
pass headers for all collection ops

ikreymer reviewed (approved):
Added some minor tweaks:
- Added resolve_relative_access_path() to ensure file URLs either use internal origin or current origin from request.
- Also made UserUploadFile be a UserFile

Really nice work! Should be good to go I think!

ikreymer committed:
further cleanup:
- consolidate UserUploadFile into SeedFile, can split again if needed
- add get_absolute_presigned_url() to UserFile for convenience in getting absolute url directly

ikreymer committed:
tests: add host header to be sure of exact host
